File: slicing-ios/SlicingGame/Page/Storage/Page.swift
//
//  Page.swift
//  Page storage: a single page item
//

import Foundation

class Page {
    
    // --
    // MARK: Members
    // --
    
    var originalData = [String: Any]()
    var creationTime: TimeInterval = 0
    private var parsedHash: String = "unknown"

    var hash: String {
        get { return parsedHash }
    }
    

    // --
    // MARK: Initialization
    // --
    
    init(jsonString: String) {
        parseString(jsonString: jsonString)
    }
    
    init(jsonData: Data) {
        parseData(jsonData: jsonData)
    }

    init(dictionary: [String: Any], hash: String = "unknown") {
        originalData = dictionary
        parsedHash = hash
        updateCreationTime()
    }


    // --
    // MARK: Parsing
    // --
    
    private func parseString(jsonString: String) {
        parseData(jsonData: jsonString.data(using: .utf8))
    }
    
    private func parseData(jsonData: Data?) {
        var resultHash = "unknown"
        originalData = [:]
        if let jsonData = jsonData {
            if let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments) {
                if let parsedData = json as? [String: Any] {
                    originalData = parsedData
                    resultHash = jsonData.sha256()
                }
            }
        }
        parsedHash = resultHash
        updateCreationTime()
    }


    // --
    // MARK: State updates
    // --
    
    func updateCreationTime() {
        creationTime = Date().timeIntervalSince1970
    }


    // --
    // MARK: Extract data
    // --
    
    var modules: [[String: Any]]? {
        get {
            return originalData["modules"] as? [[String: Any]]
        }
    }
    
    var layout: [String: Any]? {
        get {
            if let dataSets = originalData["dataSets"] as? [String: Any] {
                return dataSets["layout"] as? [String: Any]
            }
            return nil
        }
    }
    
}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/page/storage/Page.kt
package com.crescentflare.slicinggame.page.storage

import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.infrastructure.coreextensions.sha256
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Page storage: a single page item
 */
class Page {

    // --
    // Members
    // --

    var originalData = emptyMap<String, Any>()
    val hash: String
    var creationTime: Long = 0
    private val mapUtil = InflatorMapUtil()


    // --
    // Initialization
    // --

    constructor(jsonString: String) {
        var resultHash = "unknown"
        val type = object : TypeToken<Map<String, Any>>() {
        }.type
        try {
            val result = Gson().fromJson<Map<String, Any>>(jsonString, type)
            if (result != null) {
                originalData = result
                resultHash = jsonString.sha256()
            }
        } catch (ignored: Exception) {
            // No implementation
        }
        hash = resultHash
        updateCreationTime()
    }

    constructor(map: Map<String, Any>, hash: String = "unknown") {
        originalData = map
        this.hash = hash
        updateCreationTime()
    }


    // --
    // State updates
    // --

    fun updateCreationTime() {
        creationTime = System.currentTimeMillis()
    }


    // --
    // Extract data
    // --

    val layout: Map<String, Any>?
        get() {
            val dataSetMap = mapUtil.asStringObjectMap(originalData["dataSets"])
            if (dataSetMap != null) {
                val layout = mapUtil.asStringObjectMap(dataSetMap["layout"])
                if (layout != null) {
                    return layout
                }
            }
            return null
        }

}
